sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala

```diff
@@ -168,15 +168,16 @@ class CacheManager extends Logging {
       (fs, path.makeQualified(fs.getUri, fs.getWorkingDirectory))
     }
 
-    cachedData.foreach {
-      case data if data.plan.find(lookupAndRefresh(_, fs, qualifiedPath)).isDefined =>
-        val dataIndex = cachedData.indexWhere(cd => data.plan.sameResult(cd.plan))
-        if (dataIndex >= 0) {
-          data.cachedRepresentation.cachedColumnBuffers.unpersist(blocking = true)
-          cachedData.remove(dataIndex)
-        }
-        sparkSession.sharedState.cacheManager.cacheQuery(Dataset.ofRows(sparkSession, data.plan))
-      case _ => // Do Nothing
+    cachedData.filter {
```

Comment:
The problem can be shown clearly with an example code snippet:

    val t = new scala.collection.mutable.ArrayBuffer[Int]

    t += 1
    t += 2

    t.foreach { 
      case i if i > 0 =>
        println(s"i = $i")
        val index = t.indexWhere(_ == i)
        if (index >= 0) {
          t.remove(index)
        }
        println(s"t: $t")
        t += (i + 2)
        println(s"t: $t")
    }

Output:

    i = 1    // The first iteration, we get the first element "1"
    t: ArrayBuffer(2)   // "1" has been removed from the array
    t: ArrayBuffer(2, 3) // New element "3" has been inserted
    i = 3   // In next iteration, element "2" is wrongly skipped
    t: ArrayBuffer(2)     // "3" has been removed from the array
    t: ArrayBuffer(2, 5)